### DISCUSSION TITLE
Fold and input dimensions

### DISCUSSION THREAD

**Discussion by tinusd** [casadi/casadi#3333] [Q&A] [answered]
Hi there!

I am playing around with fold a bit. Like map, is there a way to "reduce" the input dimensions? Here is a little example I am trying.

Assume that I have some polynomial $f(z) = a_0 + a_1z + a_2 z^2 + a_3 z^3$. Let $a = [a_0 ~ a_1 ~ a_2  ~a_3]$. I would like to evaluate it by using Horner's rule, which in Matlab would be:

```
f = 0;
for i=N:-1:1
f = f*z + a(i)
end
```

If I would like doing this using fold, I believe I need

```
f0 = MX.sym('f0');
z = MX.sym('z');
ai = MX.sym('ai');

f = f0;
f = f*z + ai

f_func = Function('f_func',{f0,z,ai},{f});
f_fold_func = f_func.fold(N);
```
I would then get the following:
fold_f_func:(i0,i1[1x4],i2[1x4])->(o0) MXFunction

However, the input i1 (which is z) is expected to be a vector. Like map, is there any way to reduce the input to only a single value? Thanks in advance!

**Comment by jgillis** [comment, marked as answer]
Not in the present release. Fold/mapaccum is not corresponding to a primitive presently (it was pre 3.0 iirc) so its probably easy to implement in Matlab. The recipe is recursive and can be found at https://github.com/casadi/casadi/blob/e9072a208fc874622d07a664e98f46e23246ac33/casadi/core/function.cpp#L492
